rust/kernel/io_mem.rs

```diff
@@ -0,0 +1,195 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Memory-mapped IO.
+//!
+//! C header: [`include/asm-generic/io.h`](../../../../include/asm-generic/io.h)
+
+use crate::{bindings, c_types, Error, Result};
+use core::convert::TryInto;
+
+extern "C" {
+    fn rust_helper_ioremap(
+        offset: bindings::resource_size_t,
+        size: c_types::c_ulong,
+    ) -> *mut c_types::c_void;
+
+    fn rust_helper_readb(addr: *const c_types::c_void) -> u8;
+    fn rust_helper_readw(addr: *const c_types::c_void) -> u16;
+    fn rust_helper_readl(addr: *const c_types::c_void) -> u32;
+    fn rust_helper_readq(addr: *const c_types::c_void) -> u64;
+
+    fn rust_helper_writeb(value: u8, addr: *mut c_types::c_void);
+    fn rust_helper_writew(value: u16, addr: *mut c_types::c_void);
+    fn rust_helper_writel(value: u32, addr: *mut c_types::c_void);
+    fn rust_helper_writeq(value: u64, addr: *mut c_types::c_void);
+}
+
+/// Represents a memory resource.
+pub struct Resource {
+    offset: bindings::resource_size_t,
+    size: bindings::resource_size_t,
+}
+
+impl Resource {
+    pub(crate) fn new(
+        start: bindings::resource_size_t,
+        end: bindings::resource_size_t,
+    ) -> Option<Self> {
```

Comment:
Should this be `Result`? i.e. if the callers pass the wrong arguments, it is supposed to be an error, or does the notion of a "no resource" make sense for some use cases?

Reply:
It is possible for a device implementation(of a spec) not to have some optional register bank. I don't think we have enough information to decide what the appropriate error should be, so making this an `Option` we push the decision to the caller.

Reply:
I was thinking more about the `checked_*` operations, not so much about `start == 0`.

Reply:
Are you proposing that we return `Result<Option<Self>>` or `Result<Self>`?

Reply:
If `start == 0` is expected but bad arguments are not, then `Result<Option<Self>>` might be the right fit semantically, yeah.

Reply:
This is an internal API, used here: https://github.com/wedsonaf/linux/blob/pl061/rust/kernel/amba.rs#L174

I don't think making it `Result<Option<Self>>` is necessary because we don't surface this to developers, and I don't think we want to fail `probe` before even giving drivers a chance to run.